RELATED RAIL-4318: Add Web Components library

andriichumak opened the pull request:
Added a new package with WebComponents (aka Custom Elements) wrapper for Dashboard and InsightView components.

Based on #2673, please review that one first.

Still need to finish a few things:
* E2E tests setup
* Setup for manual/exploratory testing

---

Supported PR commands:

| Command                  | Description             |
| ------------------------ | ----------------------- |
| `ok to test`             | Re-run standard checks  |
| `extended test`          | BackstopJS tests        |
| `extended check sonar`   | SonarQube tests         |
| `extended check cypress` | Cypress E2E tests       |
| `extended check plugins` | Dashboard plugins tests |

---

# PR Checklist

-   [ ] commit messages adhere to the [commit message guidelines](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#what-should-the-commits-look-like)
-   [ ] review was done by a Code owner [if necessary](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-tell-if-my-pull-request-needs-approval-by-a-code-owner) (if you think it is not necessary, explain the reasoning in the description or in a comment)
-   [ ] `check` passes
-   [ ] `check-extended` passes
-   [ ] `check-extended-cypress` passes
-   [ ] `rush change` [was run if applicable](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-describe-my-changes-for-the-changelog)

gdgate commented:
Merge Failed.

Zuul merger could not merge this change into the base branch. This is most likely caused by merge conflicts. Please rebase the change and upload the rebased version. In case of further errors, contact the zuul administrator.

gdgate commented:
Merge Failed.

Zuul merger could not merge this change into the base branch. This is most likely caused by merge conflicts. Please rebase the change and upload the rebased version. In case of further errors, contact the zuul administrator.

gdgate commented:
Merge Failed.

Zuul merger could not merge this change into the base branch. This is most likely caused by merge conflicts. Please rebase the change and upload the rebased version. In case of further errors, contact the zuul administrator.

gdgate commented:
Merge Failed.

Zuul merger could not merge this change into the base branch. This is most likely caused by merge conflicts. Please rebase the change and upload the rebased version. In case of further errors, contact the zuul administrator.

gdgate commented:
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/8765/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/10602/ : SUCCESS in 18m 36s

gdgate commented:
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/8829/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/10672/ : FAILURE in 1m 03s

gdgate commented:
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/8830/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/10673/ : FAILURE in 45s

gdgate commented:
Merge Failed.

Zuul merger could not merge this change into the base branch. This is most likely caused by merge conflicts. Please rebase the change and upload the rebased version. In case of further errors, contact the zuul administrator.

gdgate commented:
Merge Failed.

Zuul merger could not merge this change into the base branch. This is most likely caused by merge conflicts. Please rebase the change and upload the rebased version. In case of further errors, contact the zuul administrator.

andriichumak commented on libs/sdk-ui-tests/.storybook/main.js:
```diff
@@ -9,4 +11,18 @@ module.exports = {
     core: {
         builder: "webpack5",
     },
+    webpackFinal: async (config) => ({
```
Copying the whole build output of the `sdk-ui-web-components` because it's using chunks in the output and Webpack does not understand chunks when used as input.

andriichumak commented on libs/sdk-ui-tests/stories/_infra/storyGroups.ts:
```diff
@@ -7,3 +7,4 @@ export const EmbeddingStories = "05 Stories For Embedding using InsightView";
 export const FilterStories = "10 Filters";
 export const ConfigurationControls = "11 Configuration Controls";
 export const UiKit = "12 UI Kit";
+export const WebComponents = "13 Web Components";
```
Web components section is rendered at the very end of the storybook. It's a bit inconsistent with the rest of the Storybook sections, that are ordered reversed (12 ... 1). Will try to find a way to order it correctly.

andriichumak commented on libs/sdk-ui-tests/stories/_infra/webComponents.ts:
```diff
@@ -0,0 +1,32 @@
+};
+
+// Configure the mock backend
+configureWebComponentsBackend(StorybookBackend(), ReferenceWorkspaceId);
```
This is an (ugly) fix that we discussed with Dan and Matyas. Needed due to the WebPack limitations on loading of the chunked libs.
The file does not export anything, sets up the mock backend for WebComponents as a side effect. This is done because the backend setup is only done once per runtime by design.

andriichumak commented on libs/sdk-ui-tests/stories/visual-regression/web-components/insightStories.tsx:
```diff
@@ -0,0 +1,51 @@
+    return <gd-insight style={{ height: 500 }} {...props} />;
+};
+
+storiesOf(`${WebComponents}/Insight`)
```
I can't run screenshots on local machine due to M1 mac. Would appreciate your help in getting screenshots for this snapshots.

no23reason replied:
Unfortunately, I also have M1 :-/
@kandl can you please help?

gdgate commented:
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/8948/ : SUCCESS in 12s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/10785/ : FAILURE in 10m 26s

andriichumak commented on libs/sdk-ui-web-components/api/sdk-ui-web-components.api.md:
```diff
@@ -0,0 +1,21 @@
+## API Report File for "@gooddata/sdk-ui-web-components"
```
I have 3 entry points for the repo - index for the main + one for each backend. This way we can load the absolute minimum in the first chunk and load the rest lazily. And user can still load tiger/bear backend in case they want custom auth setup. API extractor only works with a single entry point. I don't see it as issue, as the other two entry points are described in tiger and backend packages respectively.

no23reason replied:
I guess that is ok for now :) please add this package to common/scripts/build-docs.js to the libsToBeRemoved array so that it is not made part of the API reference.

andriichumak commented on libs/sdk-ui-web-components/src/parseUrl.ts:
```diff
@@ -0,0 +1,40 @@
+export const parseUrl = (scriptUrl: URL) => {
+    const protocol = scriptUrl.protocol.toLowerCase() === "http:" ? "http:" : "https:";
+    const host = scriptUrl.host;
+    const workspaceId = scriptUrl.pathname.match(/^\/components\/([^./]+)(\.js)?$/i)?.[1];
```
Assuming workspaceId can't include dot (.) Not sure if this is really the case or I should make this RegExp more robust.

no23reason replied:
On tiger and panther dots are allowed (just tested it), so please try to make this more robust.

gdgate commented:
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/8950/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/10787/ : FAILURE in 10m 27s

andriichumak commented on libs/sdk-ui-web-components/src/visualizations/CustomElementAdapter.ts:
```diff
@@ -0,0 +1,173 @@
+        link.type = "text/css";
+        link.rel = "stylesheet";
+        const webpackWorkaround = SHADOW_STYLES;
+        link.href = new URL(webpackWorkaround, import.meta.url).href;
```
There is a weird bug with Webpack when it's trying to interpolate `new URL('/some/path', import.meta.url)` build time. I guess it's an optimization of some sort, as they are using this pattern internally as well. Re-assigning the string value to a variable first works around that issue.

andriichumak replied:
Note, `SHADOW_STYLES` is interpolated to a string value by the `DefinePlugin`. See webpack config.

andriichumak commented on libs/sdk-ui-web-components/src/visualizations/CustomElementAdapter.ts:
```diff
@@ -0,0 +1,173 @@
+
+import { CustomElementContext, getContext } from "../context";
+// eslint-disable-next-line import/no-unassigned-import
+import "./components.shadow.css";
```
`*.shadow.css` has it's own loader that extracts the styles to a separate file, so we can load it with link tag. Regular `.css` files are simply loaded as `style` tag and auto-injected to `head`.

andriichumak commented on libs/sdk-ui-web-components/src/visualizations/components.shadow.css:
```diff
@@ -0,0 +1,13 @@
+@import "~@gooddata/sdk-ui-pivot/styles/css/main.css";
+@import "~@gooddata/sdk-ui-kit/styles/css/main.css";
+@import "~@gooddata/sdk-ui-ext/styles/css/main.css";
+@import "~@gooddata/sdk-ui-dashboard/styles/css/main.css";
```
Ideally, we would want to have 2 sets of styles - for insight and for dashboard. This is a compromise here to keep the setup simple. I tried a bunch of different style loading options and this is the optimal solution. To make it completely clean, would likely need to create a custom WebPack loader or plugin - too much work for the benefit.

Anyway, this styles are extracted to a separate highly cachable file and injected into Shadow DOMs with a link tag, so extra dashboard styles should not make a big difference even if user is only using `gd-insight`.

no23reason commented on libs/sdk-ui-web-components/LICENSE:
```diff
@@ -0,0 +1,7 @@
+Copyright (c) 2020-2021 GoodData Corporation
```
Please fix the current year.

no23reason replied:
Also make sure this correlates with the years mentioned in the README.

no23reason commented on libs/sdk-ui-web-components/README.md:
```diff
@@ -0,0 +1,130 @@
+# GoodData Web Component
+
+This library provides a set of Custom Elements (aka Web Components) to easy the embedding of GoodData visualizations
```
The formulation is a bit strange, how about
> This library provides a set of Custom Elements (aka Web Components) to make the embedding of GoodData visualizations into web applications easier.

andriichumak replied:
Yeah, it should have been "ease", not "easy" :) But I like your option better,

no23reason commented on libs/sdk-ui-web-components/README.md:
```diff
@@ -0,0 +1,130 @@
+<script type="module" src="https://{your-GD-server-URL}/components/{workspaceId}.js?auth=sso"></script>
+
+<!-- for example -->
+<script type="module" src="https://analytics.example.com/components/my-workspace.js?auth=sso"></script>
```
Please use https://example.gooddata.com as the placeholder URL, it will link to a nice page explaining what happened in case someone actually tries to use this as is. See WEB-3288 for more details.

no23reason replied:
This applies to several other places in this readme.

no23reason commented on libs/sdk-ui-web-components/README.md:
```diff
@@ -0,0 +1,130 @@
+-   `locale` - the localization of the visualization. For available languages, see [the full list of available localizations](https://github.com/gooddata/gooddata-ui-sdk/blob/master/libs/sdk-ui/src/base/localization/Locale.ts).
+-   `mapbox` - the map access token to be used by [geo pushpin charts](https://sdk.gooddata.com/gooddata-ui/docs/geo_pushpin_chart_component.html#geo-config).
+
+`gd-dashboard` emits events from [this list](https://sdk.gooddata.com/gooddata-ui-apidocs/v8.9.0/docs/sdk-ui-dashboard.dashboardeventtype.html).
```
Please use https://sdk.gooddata.com/gooddata-ui-apidocs/docs/sdk-ui-dashboard.dashboardeventtype.html so that it always points to the latest version.

no23reason commented:
extended test

no23reason commented on libs/sdk-ui-web-components/src/visualizations/Insight.tsx:
```diff
@@ -0,0 +1,59 @@
+    GET_VISUALIZATION,
+} from "./CustomElementAdapter";
+
+type IInsightView = typeof import("@gooddata/sdk-ui-ext/esm/insightView/InsightView").InsightView;
```
Can we do
```ts
import type { InsightView } from "@gooddata/sdk-ui-ext";
```
instead? That should also make the dependency cruiser checks that are currently failing pass.

no23reason commented on libs/sdk-ui-web-components/src/visualizations/Dashboard.tsx:
```diff
@@ -0,0 +1,62 @@
+    LOAD_COMPONENT,
+} from "./CustomElementAdapter";
+
+type IDashboard = typeof import("@gooddata/sdk-ui-dashboard/esm/presentation/dashboard/Dashboard").Dashboard;
```
Can we do
```ts
import type { Dashboard as IDashboard } from "@gooddata/sdk-ui-dashboard";
```
instead? That should also make the dependency cruiser checks that are currently failing pass.

andriichumak replied:
Looks like deps cruiser is worried about css imports, not this one...
https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/10805/console

no23reason commented on libs/sdk-ui-web-components/src/tests/context.test.ts:
```diff
@@ -0,0 +1,33 @@
+
+describe("context", () => {
+    // Use enforced type casting for test simplicity
+    const mockContext = { backend: "test", workspaceId: "test" } as unknown as CustomElementContext;
```
You could avoid the cast by providing a "real" backend, see https://github.com/gooddata/gooddata-ui-sdk/blob/master/libs/sdk-ui-charts/src/charts/areaChart/tests/AreaChartStacking.test.ts#L6 for example.

yenkins commented:
Link to Storybook test result: https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-backstop-tests/1738/artifact/libs/sdk-ui-tests/backstop/output/html-report/index.html

gdgate commented:
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/8959/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-backstop-tests/1738/ : SUCCESS in 27m 03s

gdgate commented:
Merge Failed.

Zuul merger could not merge this change into the base branch. This is most likely caused by merge conflicts. Please rebase the change and upload the rebased version. In case of further errors, contact the zuul administrator.

gdgate commented:
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/8962/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/10805/ : FAILURE in 10m 19s

gdgate commented:
Merge Failed.

Zuul merger could not merge this change into the base branch. This is most likely caused by merge conflicts. Please rebase the change and upload the rebased version. In case of further errors, contact the zuul administrator.

gdgate commented:
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/8963/ : SUCCESS in 5s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/10806/ : SUCCESS in 21m 25s

no23reason commented on libs/sdk-ui-web-components/src/tests/context.test.ts:
```diff
@@ -0,0 +1,33 @@
+import { dummyBackend } from "@gooddata/sdk-backend-mockingbird";
+
+describe("context", () => {
+    // Use enforced type casting for test simplicity
```
Nitpick: this comment is no longer relevant

gdgate commented:
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/8968/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/10812/ : SUCCESS in 21m 02s

andriichumak commented:
One more little improvement - I noticed that there might be a conflict between chunk file name and a workspace id. So, moved the chunks to a subfolder and changed the URL parsing algorithm a little.

gdgate commented:
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/8971/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/10816/ : SUCCESS in 24m 00s

gdgate commented:
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/8978/ : SUCCESS in 9s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/10826/ : SUCCESS in 21m 42s

gdgate commented:
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/8985/ : SUCCESS in 9s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/10832/ : SUCCESS in 20m 41s

kandl reviewed (approved):
LGTM

kandl commented:
extended test

yenkins commented:
Link to Storybook test result: https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-backstop-tests/1741/artifact/libs/sdk-ui-tests/backstop/output/html-report/index.html

gdgate commented:
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/8991/ : SUCCESS in 10s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-backstop-tests/1741/ : UNSTABLE in 29m 15s

yenkins commented:
### Sonar scan result


More detail, see in https://sonarqube-gate.intgdc.com/dashboard?id=gooddata-ui-sdk-gate-PR2735

To scan for vulnerabilities in dependencies and run unit tests (to get coverage report in sonar) please comment your PR with 'extended check sonar'.

gdgate commented:
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-fossa-scanning-zuul-el7/2567/ : SUCCESS in 9m 57s (non-voting)
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-merge-compliance-check/2866/ : SUCCESS in 20s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-sonarqube-scan-zuul-gate/1697/ : SUCCESS in 16m 23s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/10838/ : SUCCESS in 20m 04s
